mersenne-31/src/x86_64_avx512/packing.rs

```diff
@@ -270,7 +270,80 @@ fn sub(lhs: __m512i, rhs: __m512i) -> __m512i {
     unsafe {
         // Safety: If this code got compiled then AVX-512F intrinsics are available.
         let t = x86_64::_mm512_sub_epi32(lhs, rhs);
-        let u = x86_64::_mm512_add_epi32(t, P);
+        let u = x86_64::_mm512_add_epi32(t, P_AVX512);
+        x86_64::_mm512_min_epu32(t, u)
+    }
+}
+
+/// Reduce a representative in {0, ..., P^2}
+/// to a representative in [-P, P]. If the input is greater than P^2, the output will
+/// still correspond to the same class but will instead lie in [-P, 2^34].
+#[inline(always)]
+fn partial_reduce_neg(x: __m512i) -> __m512i {
+    unsafe {
+        // Get the top bits shifted down.
+        let hi = x86_64::_mm512_srli_epi64::<31>(x);
+
+        const LOW31: __m512i = unsafe { transmute::<[u64; 8], _>([0x7fffffff; 8]) };
+        // nand instead of and means this returns P - lo.
+        let neg_lo = x86_64::_mm512_andnot_si512(x, LOW31);
```

Comment:
You could also do `x86_64::_mm512_maskz_andnot_epi32(EVENS, x, P_AVX512)` to avoid introducing another constant